src/mcfret.c

```diff
@@ -0,0 +1,621 @@
+	for (i=0;i<non->singles;i++){
+	   for (j=0;j<non->singles;j++){
+	      fprintf(absorption_matrix,"%e %e ",re_S_1[t1*non->singles*non->singles+i*non->singles+j],im_S_1[t1*non->singles*non->singles+i*non->singles+j]);
+	   }
+	}
+	fprintf(absorption_matrix,"\n");
+    }
+    fclose(absorption_matrix);
+    
+    
+    /*Free the memory*/
+    free(vecr);	
+    free(veci);  
+    free(Hamil_i_e);
+    free(mu_xyz);
+}
+
+/* Sub routine for adding up the calculated response in the response function */
+void mcfret_response_function_sub(float *re_S_1,float *im_S_1,int t1,t_non *non,float *cr,float *ci){
+  int i,k;
+  int N,nn2;
+  N=non->singles;
+  nn2=N*N;
+
+  /* Update response matrix */
+  for (i=0;i<non->singles;i++){
+      for (k=0; k<non->singles; k++){
+        /* We store response function so we can do matrix multiplication */
+          re_S_1[t1*nn2+(N*i+k)]+=cr[N*i+k]; 
+          im_S_1[t1*nn2+(N*i+k)]+=ci[N*i+k];
+      }
+  }
+
+  return;
+}
+
+/* Find the average couplings but only between different segments */
+void mcfret_coupling(float *J,t_non *non){
+  /* Define variables and arrays */
+   /* Integers */
+    int nn2;
+    int N_samples;
+    int samples;
+    int ti,i,j;
+    int cl,Ncl;
+
+    /* Hamiltonian of the whole system - all donors and acceptors included */
+    float *Hamil_i_e;
+    float *mu_xyz;
+    float shift1; 
+
+  /* Time parameters */
+    time_t time_now,time_old,time_0;
+  /* Initialize time */
+    time(&time_now);
+    time(&time_0);
+    shift1=(non->max1+non->min1)/2;
+    printf("Frequency shift %f.\n",shift1);
+    non->shifte=shift1;
+
+  /* File handles */
+    FILE *H_traj;
+    FILE *C_traj;
+    FILE *mu_traj;
+    /* FILE *absorption_matrix, *emission_matrix,*/
+    FILE *log;
+    FILE *Cfile;
+
+    mu_xyz=(float *)calloc(non->singles*3,sizeof(float));
+    Hamil_i_e=(float *)calloc((non->singles+1)*non->singles/2,sizeof(float));
+
+    /* Open Trajectory files */
+    open_files(non,&H_traj,&mu_traj,&Cfile);
+
+    /* Here we want to call the routine for checking the trajectory files */ 
+    control(non);
+
+    /* Initialize sample numbers */
+    N_samples=determine_samples(non);
+    Ncl=0;
+
+    /* Read coupling, this is done if the coupling and transition-dipoles are */
+    /* time-independent and only one snapshot is stored */
+    read_coupling(non,C_traj,mu_traj,Hamil_i_e,mu_xyz);
+    samples=1;
+
+
+    log=fopen("NISE.log","a");
+    fprintf(log,"Begin sample: %d, End sample: %d.\n",non->begin,non->end);
+    fclose(log);
+
+
+    /* Looping over samples: Each sample represents a different starting point on the Hamiltonian trajectory */
+    for (samples=non->begin;samples<non->end;samples++){ 
+        ti=samples*non->sample;
+        if (non->cluster!=-1){
+            if (read_cluster(non,ti,&cl,Cfile)!=1){
+	              printf("Cluster trajectory file to short, could not fill buffer!!!\n");
+	              printf("ITIME %d\n",ti);
+	              exit(1);
```

Comment:
One should close Cfile here!

Reply:
I do not think this is really needed, if we end up here something wrong and the program stops. We also do not close all the other files and free up all the memory.